### ISSUE TITLE
API: Unhandled exception while clearing user's preferredLanguages attribute via PATCH request

### ISSUE THREAD

**Issue by mgetka** [FusionAuth/fusionauth-issues#706]
### Description 

In various parts of API documentation, there is following note about `PATCH` requests behaviour:

> When using the PATCH method, use the same request body documentation that is provided for the PUT request. The PATCH method will merge the provided request parameters into the existing object, this means all parameters are optional when using the PATCH method and you only provide the values you want changed. __A `null` value can be used to remove a value__. Patching an `Array` will result in all values from the new list being appended to the existing list, this is a known limitation to the current implementation of PATCH.

An attempt to clear `user.preferredLanguages` array via patching it with `null` causes an unhandled exception.

### Steps to reproduce
1. Try to remove `user.preferredLanguages` attribute by setting it to `null` via `PATCH` request.
2. Observe unhandled exception

### Expected behavior
The attribute is being cleared or 400 response is returned if this is a part of current API limitations.

### Platform
Docker & FusionAuth 1.17.1

### Additional context
Request
```
curl -sXPATCH 'http://localhost:9011/api/user/04497db5-9126-45de-9e0d-52af73104bc7' \
--header 'Authorization: ***' \
--header 'Content-Type: application/json' \
--data-raw '{"user": {"preferredLanguages": null}}' | jq
{
  "generalErrors": [
    {
      "code": "[Exception]",
      "message": "FusionAuth encountered an unexpected error. Please contact support for assistance."
    }
  ]
}
```
Log entries
```
Jun 26, 2020 7:32:04.458 PM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.NullPointerException: null
	at io.fusionauth.domain.User.normalize(User.java:318)
	at io.fusionauth.app.action.api.UserAction.validatePutAndPatch(UserAction.java:234)
	at jdk.internal.reflect.GeneratedMethodAccessor171.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:414)
	at org.primeframework.mvc.validation.DefaultValidationProcessor.validate(DefaultValidationProcessor.java:77)
	at org.primeframework.mvc.validation.DefaultValidationWorkflow.perform(DefaultValidationWorkflow.java:46)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.security.DefaultSecurityWorkflow.perform(DefaultSecurityWorkflow.java:81)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultPostParameterWorkflow.perform(DefaultPostParameterWorkflow.java:50)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.content.DefaultContentWorkflow.perform(DefaultContentWorkflow.java:52)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultParameterWorkflow.perform(DefaultParameterWorkflow.java:57)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.DefaultURIParameterWorkflow.perform(DefaultURIParameterWorkflow.java:102)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.scope.DefaultScopeRetrievalWorkflow.perform(DefaultScopeRetrievalWorkflow.java:58)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.message.DefaultMessageWorkflow.perform(DefaultMessageWorkflow.java:45)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.FrontEndTenantWorkflow.perform(FrontEndTenantWorkflow.java:71)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.action.DefaultActionMappingWorkflow.perform(DefaultActionMappingWorkflow.java:126)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.workflow.StaticResourceWorkflow.perform(StaticResourceWorkflow.java:97)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.parameter.RequestBodyWorkflow.perform(RequestBodyWorkflow.java:89)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at org.primeframework.mvc.security.DefaultSavedRequestWorkflow.perform(DefaultSavedRequestWorkflow.java:57)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.CORSFilter.handleNonCORS(CORSFilter.java:291)
	at io.fusionauth.app.primeframework.CORSFilter.doFilter(CORSFilter.java:252)
	at io.fusionauth.app.primeframework.CORSRequestWorkflow.perform(CORSRequestWorkflow.java:48)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:43)
	at io.fusionauth.app.primeframework.FusionAuthMVCWorkflow.perform(FusionAuthMVCWorkflow.java:88)
	at org.primeframework.mvc.workflow.DefaultWorkflowChain.continueWorkflow(DefaultWorkflowChain.java:44)
	at org.primeframework.mvc.servlet.FilterWorkflowChain.continueWorkflow(FilterWorkflowChain.java:50)
	at org.primeframework.mvc.servlet.PrimeFilter.doFilter(PrimeFilter.java:82)
	at com.inversoft.maintenance.servlet.MaintenanceModePrimeFilter.doFilter(MaintenanceModePrimeFilter.java:59)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.inversoft.servlet.UTF8Filter.doFilter(UTF8Filter.java:27)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:543)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:615)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1623)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:832)

```

## Related bugs

- Also filed here: https://github.com/FusionAuth/fusionauth-issues/issues/956
- `Null` causes problem JSON, not currently supported, but somewhat related: https://github.com/FusionAuth/fusionauth-issues/issues/254

**Comment by mooreds** [comment]
Nice find!

**Comment by nicko** [comment]
Hi, `1b1f705` looks like a pretty low-risk fix and it's been open since July. Any chance you can merge it soon? We actually can't proceeed with using the language setting feature in Fusion at the moment.

**Comment by robotdan** [comment]
@nicko you should be able to use `PUT` in conjunction with a `GET` as a work around. Is this not working for you?

**Comment by robotdan** [comment]
Would prefer to fix this under https://github.com/FusionAuth/fusionauth-issues/issues/441 when adding support for RFC 7396.

However, in theory we could add a temporary fix for all of our arrays using the fix proposed by @bguyza via https://github.com/FusionAuth/fusionauth-java-client/pull/4/files.

**Comment by mgetka** [comment]
I'm not so familiar with Java, so I'am not sure whether I'm understanding this fix correctly. Would it just silently strip all array assignments to `null` in `PATCH` requests? If it is so, this will most likely mislead the developers into thinking that everything went good, when clearly it didn't. Encountering an exception shows me that the thing I wanted to accomplish did not succeed, and I can look for a workaround (eg. based on a discussion in this issue).

**Comment by fabiosimeoni** [comment]
> Would prefer to fix this under #441 when adding support for RFC 7396.
> 
> However, in theory we could add a temporary fix for all of our arrays using the fix proposed by @bguyza via https://github.com/FusionAuth/fusionauth-java-client/pull/4/files.

Does this mean one can no longer reset the `preferredLanguages` (`null` interpreted as empty array)? I'm seeing my clumsy attempts to send a `preferredLanguages:null` return with a `200` and absolute no changes to the field, like with the append-semantics of empty arrays.

If you confirm this, we're in pretty tight spot now:

- cannot overwrite arrays with `PATCH` requests until #441 is delivered.
-  cannot follow the temporary workaround suggested in #441 of issuing two `PATCH`-requests (which was already bad due to the loss of atomicity).

I wouldn't really want to give up `PATCH` and move to `PUT`, as that would either bind my code to the full `User` model and its evolutions (am really using 4 fields). Or else refactor everything and accept a dependency to your client library, which is overkill for me.

If I'm missing an easy solution to avoid the append semantics, I'd be very grateful if you could share it.

**Comment by robotdan** [comment]
Closing. In version `1.39.0`, new `Content-Type` values of `application/json-patch+json` or `application/merge-patch+json` will be available which should allow you to manage `PATCH` much better. 
 
- https://github.com/FusionAuth/fusionauth-issues/issues/441
